ARCs/arc-0020.md

```diff
@@ -0,0 +1,870 @@
+>     supply of the Smart ASA.
+>   - `name` and `unit_name` MAY be set to `SMART-ASA` and `S-ASA`, to
+>     denote that this ASA is Smart and has a controlling application.
+> - Persist the `total`, `decimals`, `default_frozen`, etc. fields for later
+>   use/retrieval.
+> - Return the ID of the created ASA.
+>
+> It is RECOMMENDED for calls to this method to be permissioned, e.g. to only
+> approve transactions issued by the controlling Smart Contract creator.
+
+#### Asset Configuration
+
+```json
+{
+  "name": "asset_config",
+  "args": [
+    { "type": "asset", "name": "config_asset" },
+    { "type": "uint64", "name": "total" },
+    { "type": "uint32", "name": "decimals" },
+    { "type": "bool", "name": "default_frozen" },
+    { "type": "string", "name": "unit_name" },
+    { "type": "string", "name": "name" },
+    { "type": "string", "name": "url" },
+    { "type": "byte[]", "name": "metadata_hash" },
+    { "type": "address", "name": "manager_addr" },
+    { "type": "address", "name": "reserve_addr" },
+    { "type": "address", "name": "freeze_addr" },
+    { "type": "address", "name": "clawback_addr" }
+  ],
+  "returns": { "type": "void" }
+}
+```
+
+Calling `asset_config` configures an existing Smart ASA.
+
+> Upon a call to `asset_config`, a reference implementation SHOULD:
+>
+> - Fail if `config_asset` does not correspond to an ASA controlled by this smart
+>   contract.
+> - Succeed iff the `sender` of the transaction corresponds to the `manager_addr`
+>   that was previously persisted for `config_asset` by a previous call to this
+>   method or, if never caller, to `asset_create`.
+> - Update the persisted `total`, `decimals`, `default_frozen`, etc. fields for later
+>   use/retrieval.
+>
+> The business logic associated to the update of the other parameters is left to
+> the implementation. An implementation that maximizes similarities with ASAs,
+> SHOULD NOT allow modifying the `clawback_addr` or `freeze_addr` after they
+> have been set to the special value `ZeroAddress`.
+>
+> The implementation MAY provide flexibility on the fields of an ASA that
+> cannot be updated after initial configuration. For instance, it MAY update the
+> `total` parameter to enable minting of new units or restricting the maximum
+> supply; when doing so, the implementation SHOULD ensure that the updated
+> `total` is not lower than the current circulating supply of the asset.
+
+#### Asset Transfer
+
+```json
+{
+  "name": "asset_transfer",
+  "args": [
+    { "type": "asset", "name": "xfer_asset" },
+    { "type": "uint64", "name": "asset_amount" },
+    { "type": "account", "name": "asset_sender" },
+    { "type": "account", "name": "asset_receiver" }
+  ],
+  "returns": { "type": "void" }
+}
+```
+
+Calling `asset_transfer` transfers a Smart ASA.
+
+> Upon a call to `asset_transfer`, a reference implementation SHOULD:
+>
+> - Fail if `xfer_asset` does not correspond to an ASA controlled by this smart
```

Comment:
Should there be mention here of looking up the frozen statuses set by `asset_freeze` and `account_freeze`?